source/Halibut/Transport/Streams/TimeoutStreamWrapper.cs

```diff
@@ -0,0 +1,192 @@
+
+                return await actionTask;
+            }
+            catch (Exception e)
+            {
+                ThrowMeaningfulException(e);
+
+                throw;
+            }
+
+            void ThrowMeaningfulException(Exception? innerException = null)
+            {
+                if (timeoutCancellationTokenSource.IsCancellationRequested)
+                {
+                    throw new OperationCanceledException($"The {methodName} operation timed out after {TimeSpan.FromMilliseconds(timeout)}.", innerException);
+                }
+
+                if (cancellationToken.IsCancellationRequested)
+                {
+                    throw new OperationCanceledException($"The {methodName} operation was cancelled.", innerException);
+                }
+            }
+        }
+
+        public override void Close() => inner.Close();
+
+        public override async Task CopyToAsync(Stream destination, int bufferSize, CancellationToken cancellationToken) => await inner.CopyToAsync(destination, bufferSize, cancellationToken).ConfigureAwait(false);
+
+        public override int EndRead(IAsyncResult asyncResult) => inner.EndRead(asyncResult);
+
+        public override void EndWrite(IAsyncResult asyncResult) => inner.EndWrite(asyncResult);
+
+        public override int ReadByte() => inner.ReadByte();
+
+        public override IAsyncResult BeginRead(byte[] buffer, int offset, int count, AsyncCallback? callback, object? state) => inner.BeginRead(buffer, offset, count, callback, state);
+
+        public override IAsyncResult BeginWrite(byte[] buffer, int offset, int count, AsyncCallback? callback, object? state) => inner.BeginWrite(buffer, offset, count, callback, state);
+
+        public override void Flush() => inner.Flush();
+
+        public override int Read(byte[] buffer, int offset, int count) => inner.Read(buffer, offset, count);
+
+        public override long Seek(long offset, SeekOrigin origin) => inner.Seek(offset, origin);
+
+        public override void SetLength(long value) => inner.SetLength(value);
+
+        public override void Write(byte[] buffer, int offset, int count) => inner.Write(buffer, offset, count);
+
+        public override void WriteByte(byte value) => inner.WriteByte(value);
+
+#if !NETFRAMEWORK
+        public override void CopyTo(Stream destination, int bufferSize) => inner.CopyTo(destination, bufferSize);
+        
+        public override async ValueTask DisposeAsync() => await inner.DisposeAsync().ConfigureAwait(false);
+        
+        public override int Read(Span<byte> buffer) => inner.Read(buffer);
+
+        public override void Write(ReadOnlySpan<byte> buffer) => inner.Write(buffer);
+#endif
+
+#if NETFRAMEWORK
+        public override ObjRef CreateObjRef(Type requestedType) => inner.CreateObjRef(requestedType);
+
+        public override object? InitializeLifetimeService() => inner.InitializeLifetimeService();
+#endif
+
+        public override int ReadTimeout
```

Comment:
Just wondering, why did we go with an int instead of a TimeSpan?

If we stay with the int, did we want to rename this ReadTimeoutMilliseconds to be explicit?

Reply:
This overrides the Stream properties, out of our control.

Reply:
Ah, of course 🤦